### DISCUSSION TITLE
Self-reference dependency has unused export name: This should not happen

### DISCUSSION THREAD

**Discussion by framerate** [webpack/webpack#13988] [Q&A]
# Bug report

I've been trying to port our webpack 4 app over to v5 and everything went great in development. But when I tried to fix our production build, I get a bunch of these (many from the  `aws-iot-device-sdk-v2` and `lodash` packages. 

The reason I'm flagging this as a bug report is because there is currently no way I can find to figure this out. It COULD be user error (but since its all coming from `node_modules` I tend to think not) but without this issue/error being documented (or at least I couldn't find it) I'm unsure how to proceed. 

Can someone point me in the proper direction of how to go about debugging/fixing this? #11673 is the only bug I can find and that was fixed in 5.11, I'm running 5.50 right now so I'm at a loss.

```
ERROR in ./node_modules/aws-crt/dist.browser/browser.js
Self-reference dependency has unused export name: This should not happen
Error: Self-reference dependency has unused export name: This should not happen
    at CommonJsSelfReferenceDependencyTemplate.apply (/home/justin/code/openpath/Platinum/node_modules/webpack/lib/dependencies/CommonJsSelfReferenceDependency.js:102:10)
    at JavascriptGenerator.sourceDependency (/home/justin/code/openpath/Platinum/node_modules/webpack/lib/javascript/JavascriptGenerator.js:205:12)
    at JavascriptGenerator.sourceModule (/home/justin/code/openpath/Platinum/node_modules/webpack/lib/javascript/JavascriptGenerator.js:112:9)
    at JavascriptGenerator.generate (/home/justin/code/openpath/Platinum/node_modules/webpack/lib/javascript/JavascriptGenerator.js:98:8)
    at NormalModule.codeGeneration (/home/justin/code/openpath/Platinum/node_modules/webpack/lib/NormalModule.js:1144:22)
    at /home/justin/code/openpath/Platinum/node_modules/webpack/lib/Compilation.js:2732:22
    at /home/justin/code/openpath/Platinum/node_modules/webpack/lib/Cache.js:93:5
    at Hook.eval [as callAsync] (eval at create (/home/justin/code/openpath/Platinum/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Cache.get (/home/justin/code/openpath/Platinum/node_modules/webpack/lib/Cache.js:75:18)
    at ItemCacheFacade.get (/home/justin/code/openpath/Platinum/node_modules/webpack/lib/CacheFacade.js:117:15)
 @ ./node_modules/aws-iot-device-sdk-v2/dist/index.js 38:18-36
 @ ./app/global/mqtt/sagas.js 41:0-40 204:19-29
 @ ./app/routes/AuthenticatedContainer/route.js
 @ ./app/routes/AppContainer/index.js 55:0-69 177:290-308
 @ ./app/app.prod.js 26:0-38 153:15-18
 @ ./app/app.js
```

**What is the current behavior?**
The build for production fails with `Self-reference dependency has unused export name: This should not happen` from some npm modules and I'm not sure how to diagnose

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**
I would expect the build to work in production since it worked with a very similar development build.

**Other relevant information:**
webpack version:  webpack 5.50.0
Node.js version:  v14.16.1
Operating System: Windows 11 (WSL2 Ubuntu 20.04)
Additional tools:

Full log available here: https://gist.github.com/framerate/a9bc07caef3c6c5e842ad3417b25398a

**Comment by alexander-akait** [comment]
> If the current behavior is a bug, please provide the steps to reproduce.

Can you provide small example?

**Reply by framerate** [reply]
I am trying. Our project is a large corporate application so the webpack file is rather large. I was completely blocked by this error and had to move on but I want to return to the branch next week.

Can you give me any info on this error or perhaps some common reasons it would appear? Maybe in reference to all of them being in my "node_modules" folder. Could that be indicative of something I did wrong?

**Reply by alexander-akait** [reply]
Can you provide `resolve` options (https://webpack.js.org/configuration/resolve/)?

**Reply by framerate** [reply]
Our development build looks to be (seems to work great)
  ```
    resolve: {
      modules: ['node_modules', 'app'],
      symlinks: false,
    },
  ```
  
  Production (produces the error) looks to be:
  ```
    resolve: {
      modules: ['node_modules', 'app']
    },
  ```

**Reply by alexander-akait** [reply]
> ERROR in main.f6af3212b0c353c8efb9.js from Terser
> Invalid assignment [main.f6af3212b0c353c8efb9.js:2803,27]

something generate invalid code, maybe related, need to look at loaders, do you use babel?

**Reply by framerate** [reply]
Yes, we have a babel set up as well. 

All of the errors I'm getting relate to files in node_modules. Resolve should still be set to include node_modules, right?

**Reply by alexander-akait** [reply]
You can use this `resilve.modules: ['...', 'app']`, but it should be not a problem, I am afraid something break module code and you got this problem, without configuration(s) hard to say who...

**Reply by alexander-akait** [reply]
Can you try to enable `devtool: 'source-map'` and should the error from `terser-webpack-plugin`?

**Reply by framerate** [reply]
I tried `devtool: 'source-map'` and same problem. Final two errors look like this:

```
ERROR in ./node_modules/websocket-stream/node_modules/readable-stream/lib/internal/streams/BufferList.js
Self-reference dependency has unused export name: This should not happen
 @ ./node_modules/websocket-stream/node_modules/readable-stream/lib/_stream_readable.js 81:17-57
 @ ./node_modules/websocket-stream/node_modules/readable-stream/readable-browser.js 1:10-63
 @ ./node_modules/websocket-stream/stream.js 3:16-52
 @ ./node_modules/aws-iot-device-sdk/device/lib/ws.js 19:18-45
 @ ./node_modules/aws-iot-device-sdk/device/index.js 550:19-38
 @ ./node_modules/aws-iot-device-sdk/index.js 19:0-43
 @ ./app/global/mqtt/sagas.js 41:0-37 204:19-29
 @ ./app/routes/AuthenticatedContainer/route.js
 @ ./app/routes/AppContainer/index.js 55:0-69 177:290-308
 @ ./app/app.prod.js 26:0-38 153:15-18
 @ ./app/app.js

ERROR in main.f8b2cd3f78077133e298.js
main.f8b2cd3f78077133e298.js from Terser plugin
Invalid assignment [main.f8b2cd3f78077133e298.js:2803,27]
    at js_error (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:550:11)
    at croak (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:1274:9)
    at maybe_assign (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:3443:13)
    at expression (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:3452:24)
    at simple_statement (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:1583:55)
    at statement (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:1396:19)
    at _embed_tokens_wrapper (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:1339:26)
    at block_ (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:2172:20)
    at _function_body (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:2084:21)
    at arrow_function (/Users/justin/code/openpath/webpack-platinum/node_modules/terser-webpack-plugin/node_modules/terser/dist/bundle.min.js:1692:20)
```

Also my `resolve.modules` already includes `app` as the last entry... if that's what you meant.

Babel Config:
```{
  "presets": [
    [
      "@babel/preset-env",
      {
        "modules": false,
        "targets": {
          "browsers": [
            "last 2 versions",
            "ie >= 11"
          ]
        }
      }
    ],
    "@babel/preset-react"
  ],
  "plugins": [
    "@babel/plugin-syntax-dynamic-import",
    [
      "@babel/plugin-proposal-decorators",
      {
        "legacy": true
      }
    ],
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-optional-chaining"
  ],
  "env": {
    "production": {
      "only": [
        "app",
        "cloudUnlockApp"
      ],
      "plugins": [
        "transform-react-remove-prop-types",
        "@babel/plugin-transform-react-inline-elements",
        "@babel/plugin-transform-react-constant-elements"
      ]
    },
    "development": {
      "plugins": [
        [
          "@simbathesailor/babel-plugin-use-what-changed",
          {
            "active": true
          }
        ]
      ],
      "presets": [
        [
          "@babel/preset-react",
          {
            "development": true
          }
        ]
      ]
    },
    "test": {
      "plugins": [
        "@babel/plugin-transform-modules-commonjs",
        "@babel/plugin-syntax-dynamic-import"
      ]
    }
  }
}
```

**Comment by framerate** [comment]
Also important to note, maybe, I'm not using TerserPlugin directly anywhere, so its likely something internal.

**Comment by framerate** [comment]
It appears to build with 

```
    optimization: {
      sideEffects: false,
    },
```
    
If that helps you help me understand!
    
Possibly related to this again: https://github.com/webpack/webpack/issues/11673

**Reply by alexander-akait** [reply]
Can you provide steps to reproduce, you should not disable `sideEffects`

**Reply by framerate** [reply]
I don't believe I can unfortunately, which is a bummer for both of us. My job keeps me extremely busy and I have a side job after hours as well.

I can try to make a skeleton project though and see if it reproduces without any proprietary code. 

Can you explain to me what `sideEffects` is doing and why I shouldn't disable it (and why the other bug says to do it to fix a very similar issue?)

Any info would be appreciated.

**Reply by alexander-akait** [reply]
`sideEffects` is means functions in these files do not have side effects, i.e. changing DOM/styles/title/etc (pure functions) https://webpack.js.org/guides/tree-shaking/
